### PR TITLE
Revert workaround for old SyntaxGenerator bug

### LINKED ISSUE #4717: Remove workaround and make fixer language-agnostic
https://github.com/dotnet/roslyn-analyzers/blob/b148fa89c974a72c28184931767835709f34d1fa/src/NetAnalyzers/CSharp/Microsoft.CodeQuality.Analyzers/Maintainability/CSharpUseNameofInPlaceOfString.Fixer.cs#L18-L28

The SyntaxGenerator bug was fixed in https://github.com/dotnet/roslyn/pull/42214

### PR DESCRIPTION
**Pull request by Youssef1313**
Fixes #4717

### CODE REVIEW

File: src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/Maintainability/UseNameofInPlaceOfString.Fixer.cs
Comment (Youssef1313) [inline src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/Maintainability/UseNameofInPlaceOfString.Fixer.cs:57]: The usage of explicit type generates an IDE0007 warning locally. I'm completely confused how this doesn't fail CI?

Comment (Youssef1313) [reply src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/Maintainability/UseNameofInPlaceOfString.Fixer.cs:57]: Actually, there are **tons** of warnings locally. Including RS1024 which is explicitly disabled in editorconfig

image

image

**Review by Evangelink — APPROVED** [review]

**Comment by sharwell** [comment]
Currently the analyzers support Roslyn 3.3.1 and newer. The `SyntaxGenerator` bug was only fixed for 3.6.0.

Seems like this change needs to be blocked on a future update (unknown time) when this version is updated to 3.6.0 or newer:

https://github.com/dotnet/roslyn-analyzers/blob/33582d0369adb7657731b611f4811e981dcb2f7f/eng/Versions.props#L31

**Comment by mavasani** [comment]
> Seems like this change needs to be blocked on a future update (unknown time) when this version is updated to 3.6.0 or newer:

@sharwell There is no API dependency in the fix here that requires us to update to newer Microsoft.CodeAnalysis package. I feel we should do the following:

1. Take this change as-is
2. If user installs this package and uses the code fix on a VS version before VS2019 16.6, they will hit the original bug. We would treat this as a VS bug which has been fixed and request customer to move to VS2019 16.6 or later. IMO, that is a reasonable story.

**Review by mavasani — APPROVED** [review]
LGTM, but will wait for @sharwell's approval.

**Comment by sharwell** [comment]
@mavasani if a user hits this and reports it, the issue is going to get routed to you anyway 😄

**Comment by Youssef1313** [comment]
@sharwell Users can never hit this bug, even with the buggy behavior of SyntaxGenerator. The IDE will re-parse the text produced anyway. This workaround was mainly done for the tests here to pass, as the tests don't reparse the syntax tree.

**Comment by mavasani** [comment]
@Youssef1313 - Ah yes, thanks for the clarification. I recall the issue now, and the fact that it was test-only workaround. This gives me lot more confidence to go ahead with the change.

**Comment by sharwell** [comment]
> This workaround was mainly done for the tests here to pass, as the tests don't reparse the syntax tree.

The tests always reparse the syntax tree:

https://github.com/dotnet/roslyn-sdk/blob/6f9cf96ec9e94453ac750c50c3b84c554037edcb/src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/CodeActionTest%601.cs#L182

Sounds like the change that made this PR viable was switching to the new test library a while back.

**Comment by Youssef1313** [comment]
> The tests always reparse the syntax tree:

@sharwell Thanks for clarifying. It should the old tests which didn't reparse.

**Pull request merged**
